client/my-sites/stats/utils.js

```diff
@@ -47,3 +47,18 @@ export const trackStatsAnalyticsEvent = ( eventName, properties = {} ) => {
 	const event_from = isOdysseyStats ? 'jetpack_odyssey' : 'calypso';
 	recordTracksEvent( `${ event_from }_${ eventName }`, properties );
 };
+
+/**
+ * Prepare query string for redirection on Odyssey Stats
+ * Query string from window.location.search differs depending on environment.
+ * Make sure to append the query if we are working inside wp-admin otherwise it will be lost.
+ * @param {string} baseUrl base URL with window.location.search for Calypso Stats
+ * @param {boolean} isOdyssey is Stats running in wp-admin
+ * @param {Object} query query object
+ * @returns pathname concatenated with query string
+ */
+export const parseQueryStringForOdysseyRedirect = ( baseUrl, isOdyssey, query = {} ) => {
+	const queryString = isOdyssey && query ? `&${ new URLSearchParams( query ).toString() }` : '';
+
+	return `${ baseUrl }${ queryString }`;
+};
```

Comment:
I tested the following and it's working in my testing.

```
export const parseQueryStringForOdysseyRedirect = ( baseUrl, isOdyssey, query = {} ) => {
	const newParams = new URLSearchParams( window.location.search );
	if ( isOdyssey ) {
		const queryParams = new URLSearchParams( query );
		queryParams.forEach( ( value, key ) => {
			newParams.set( key, value );
		} );
	}

	return `${ baseUrl }?${ newParams.toString() }`;
};
```

We'd then need to upate the two call sites to not include `window.location.search` in the base URL value.

What do you think?

Reply:
I would not prefer to make the function process inputs along with a global variable simultaneously, which would make it harder to understand.

Reply:
Yes, good point. We could make the `window.location.search` a passed-in value to avoid that. BUT… @kangzj 's comment [below](https://github.com/Automattic/wp-calypso/pull/97658#discussion_r1896190198) about not using `window.location.search` at all probably makes more sense.